### ISSUE TITLE
Android app for browser-phone

### ISSUE THREAD

**Issue by vieridipaola** [InnovateAsterisk/Browser-Phone#110]
Hi,

I know there might be some work going on with Cordova or Electron, but to say the truth, I think Browser-Phone works OK with a smartphone/tablet just the way it is (sure, improvements are welcome). 

I'm wondering however how one can go about adding Browser-Phone to, say, an Android device so that it can act just like any other app. In particular, I'd like it to automatically start at boot time, and have it running all the time, even when the user opens other apps, with incoming call notifications, etc.

Wouldn't a simple and quick solution be an Android webview?

For instance, one could create a simple app with just a webview which would load the Browser-Phone URL:
 
https://developer.android.com/guide/webapps/webview

However, I don't know if "notifications" would work properly (for incoming calls).
If the webview doesn't handle notifications as a full-blown browser does, maybe a "service" would help

**Comment by InnovateAsterisk** [comment]
> I'm wondering however how one can go about adding Browser-Phone to, say, an Android device so that it can act just like any other app. In particular, I'd like it to automatically start at boot time, and have it running all the time, even when the user opens other apps, with incoming call notifications, etc.

I'm not a mobile expert, but from what i understand this drains the battery significantly - the best alternative is to use push notification, or a combination of background app refresh.

> Wouldn't a simple and quick solution be an Android webview?

I believe yes, this is correct, however to make it work correctly, some form of notifications would be required. Call Kit would be required for using the app while locked (iPhone). I also understand Cordova is the "link" between phone features and webview.

As soon as the XMPP implementation is complete, ill get onto this.

**Comment by vieridipaola** [comment]
Cordova... OK, I hope so because I've been trying out a simple webview app and all seems to be working OK except:

1) I cannot set the earpiece as audio output

2) I cannot "lock" the screen (WakeLock) when approaching the device to the ear

3) I know how to "link" javascript code from phone.js to the Android app so it can, say, "make a Toast message" pop up when there's an incoming call, but I don't know how others (such as whatsapp) have implemented a way to make sure that all incoming calls or messages are notified even if the app is not in the foreground. Maybe there has to be a service running and launched at boot time that actually has already all the SIP credentials to put the client "on-line" so that each time there's a new call, a notification is shown so the main GUI can be restored to the foreground.

So yes, it's not as simple as it sounds but using a webview might speed up the process.

**Comment by vieridipaola** [comment]
BTW I know Signal (https://github.com/signalapp/) does not use SIP or XMPP -- even though it might be an extension -- (and for now I'd rather use open-source SIP and XMPP server-side implementations like Asterisk, Prosody or ejabberd) but maybe the client source codes can be useful to the Browser-Phone project (or at least inspiring).

**Comment by InnovateAsterisk** [comment]
I have made quite a bit of progress on the XMPP front. It's a reasonable protocol, and solves most of the issues. I'm using Openfire to test, and the strophe.js library. As much as possible sticking to the standards track... so it should work on any XMPP server.

**Comment by vieridipaola** [comment]
Great! Can't wait to try it out!

**Comment by vieridipaola** [comment]
Hi Conrad,

The XMPP code is already working albeit not all features are implemented.
I was wondering if you could give an update as to how the "mobile version" is going.
Browser-Phone is so close to be used on a smartphone yet there are a few things to be addressed -- and they are challenging.
Even for basic use without push messages and all, I still can't seem to set the earpiece as audio output and "lock" the screen (WakeLock) when approaching the device to the ear.
That alone would be enough to start using Browser-Phone on any portable device.

**Comment by InnovateAsterisk** [comment]
Yes, the XMPP does work, but you must set it up quite specifically - it will need a walkthrough. And yes, there are limited features for now.

With regards to mobile... I have taken an in-depth look at the mobile options for this project and discovered a number of very interesting things. Unfortunately is has taken up a significant amount of time, but I was not able to come up with the prefect plan, and has created a delay in the XMPP and video conferencing development.

Over the last few weeks I have upgraded the SIPJS library to 0.20.0. This is a huge change especially to the configuration of the SIP eventing. With this library i'm now able to fully accommodate for SFU in the Asterisk ConfBridge application, and includes things like participant names, talker detection, etc. With some clever dialplan work, and a hosting page that will support it, one could develop a Zoom/Meet like solution. 

(_Note: Asterisk is limited to 16 video streams - though it must be said, it may be better to change your strategy to "talker detection" or "marked user" to hold a useful conversation with more than 16 people._)

With regards to Mobile: I was able to run the Browser phone on both Safari on iOS, and Chrome on Android. They both worked, but with varying degrees of success... However since you cannot develop any of your own features like proximity etc - this approach comes to an end quickly. (Note: if none of the mobile features like push etc are required, then you can use it just like this.)

 I can run the page via a simple test app on both Android and iOS using a "WebView", however neither of them worked out the box (not surprisingly), there are a number of permissions and settings to adjust. I hunted to find settings that would allow ```getUserMedia()```, but it appears it may simply be a feature Apple are simply not willing to allow just yet. 

Part of my research lead me to discover that WebRTC is in fact an open source library, written in c++, and easily available, so this means all that one would need is a SIP & WebSocket library to wrap the WebRTC code with, and you would have a fully fledged mobile phone using WebRTC. There are some libraries that offer this, but it seems they form part of a larger system of commercial software, so although you can use them free, you can only connect via their commercial voice systems. (it's probably their business model - fair enough)

I still think the best approach so far is for the VebView to host the page with SipJS, and perform the SIP stack using the normal WebRTC code that comes with the WebView, and have the native UI for the buttons, settings etc. I just done know how to achieve this on both Android and iOS. (This does appear to be possible on Android)

**Comment by vieridipaola** [comment]
Hi,
Is the limit of 16 streams the same as the one described in this blog? https://www.asterisk.org/asterisk-15-multi-stream-media-sfu/

It seems to be a pjsip hard limit in the source code. I don't know if there's a reason for that.

Running Browser-Phone directly in a browser on smartphones has two basic features that do not work /at least in my case): 
1 - send audio out to the earpiece (users usually do not want everyone to hear what the other party has to say)
2 - disable screen so accidentally touching it while holding it close to your ear does not trigger an unwanted event (like calling another app and cutting off Browser-Phone)

I've done this on Android smartphones. I haven't  tried it on Apple devices.

I find these 2 items to be important in order the use Browser-Phone as-is in a smatphone's browser or webview.

**Comment by InnovateAsterisk** [comment]
> Is the limit of 16 streams the same as the one described in this blog? https://www.asterisk.org/asterisk-15-multi-stream-media-sfu/

Correct. 

> It seems to be a pjsip hard limit in the source code. I don't know if there's a reason for that.

It also seems like its a variable, that you can change in the source code, but then you will probably need to re-compile. 

> Running Browser-Phone directly in a browser on smartphones has two basic features that do not work /at least in my case):
1 - send audio out to the earpiece (users usually do not want everyone to hear what the other party has to say)
2 - disable screen so accidentally touching it while holding it close to your ear does not trigger an unwanted event (like calling another app and cutting off Browser-Phone)

Agreed. I would also add, that when the screen is off, or the app is in the background, no communications are possible, so the extension de-registers. All these considered, this app will have to run in a webview (and probably build with native UI, at least for some things) in order to be useful on a mobile device.

**Comment by rtjoa26** [comment]
Hi,

The variable limit can be found at /asterisk-..../third-party/pjproject/patches/config_site.h
#define	PJMEDIA_MAX_SDP_MEDIA   16

The PJMEDIA_MAX_SDP_MEDIA macro defines maximum SDP media lines in a SDP session descriptor.
https://www.pjsip.org/pjmedia/docs/html/group__PJMEDIA__SDP.htm#ga8ef8ceda820f28c63715192c5386dc80

You need to re-compile it.

**Comment by vieridipaola** [comment]
About the earpiece maybe this can be helpful:

https://stackoverflow.com/questions/69336155/android-webview-play-audio-through-earpiece-or-bluetooth/69386868#69386868

or maybe this if using Cordova:

https://github.com/cordova-rtc/cordova-plugin-iosrtc/blob/master/docs/iosrtc.md#iosrtcselectaudiooutputoutput

**Comment by InnovateAsterisk** [comment]
Whatever we tried, we got some sort of gotcha, that’s is until about 3 weeks ago!!

We finally cracked it! The next version of Siperb will be with beautifully crafted Tailwind CSS, in web view (the exact same html and JavaScript we use in web). Microphone working from call kit, no problem!! 🎉
